lgc/patch/PatchCopyShader.cpp

```diff
@@ -219,51 +260,73 @@ bool PatchCopyShader::runOnModule(Module &module) {
   }
 
   if (outputStreamCount > 1 && resUsage->inOutUsage.enableXfb) {
-    // StreamId = streamInfo[25:24]
-    auto streamInfo = getFunctionArgument(entryPoint, CopyShaderUserSgprIdxStreamInfo);
-
-    Value *streamId = builder.CreateIntrinsic(Intrinsic::amdgcn_ubfe, builder.getInt32Ty(),
-                                              {
-                                                  streamInfo,
-                                                  builder.getInt32(24),
-                                                  builder.getInt32(2),
-                                              });
-
-    //
-    // copyShader() {
-    //   ...
-    //   switch(streamId) {
-    //   case 0:
-    //     export outputs of stream 0
-    //     break
-    //   ...
-    //   case rasterStream:
-    //     export outputs of raster stream
-    //     break
-    //   ...
-    //   case 3:
-    //     export outputs of stream 3
-    //     break
-    //   }
-    //
-    //   return
-    // }
-    //
-
-    // Add switchInst to entry block
-    auto switchInst = builder.CreateSwitch(streamId, endBlock, outputStreamCount);
-
-    for (unsigned streamId = 0; streamId < MaxGsStreams; ++streamId) {
-      if (resUsage->inOutUsage.gs.outLocCount[streamId] > 0) {
-        std::string blockName = ".stream" + std::to_string(streamId);
-        BasicBlock *streamBlock = BasicBlock::Create(*m_context, blockName, entryPoint, endBlock);
-        builder.SetInsertPoint(streamBlock);
-
-        switchInst->addCase(builder.getInt32(streamId), streamBlock);
-
-        exportOutput(streamId, builder);
-        builder.CreateBr(endBlock);
+    if (!m_pipelineState->getNggControl()->enableNgg) {
+      // StreamId = streamInfo[25:24]
+      auto streamInfo = getFunctionArgument(entryPoint, CopyShaderUserSgprIdxStreamInfo);
+
+      Value *streamId = builder.CreateIntrinsic(Intrinsic::amdgcn_ubfe, builder.getInt32Ty(),
+                                                {
+                                                    streamInfo,
+                                                    builder.getInt32(24),
```

Comment:
I suppose 24 is the bit offset of some HW register. it is better to a constant variable for it. since it may different among different gfxip.

Reply:
Yes, it is the first system value SGPR with mixed stream info and [25:24] is the stream ID. I will add a comment here and all GPU generations from GFX6 - GFX10 have the consistent HW definitions without any variation.